src/hooks.ts

```diff
@@ -36,3 +42,112 @@ export const useHasResourcePermission = (resource: Resource, authzUrl: string, p
     const { permissions, isFetching } = useResourcePermissions(resource, authzUrl) ?? {};
     return { isFetching, hasPermission: permissions.includes(permission) };
 };
+
+export const useOpenIdConfig = (openIdConfigUrl: string) => {
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        dispatch(fetchOpenIdConfiguration(openIdConfigUrl));
+    }, [dispatch, openIdConfigUrl]);
+
+    return useSelector((state: RootState) => state.openIdConfiguration.data);
+};
+
+export const useSignInPopupTokenHandoff = (
+    applicationUrl: string,
+    authCallbackUrl: string,
+    clientId: string,
+    windowMessageHandler: MutableRefObject<null | MessageHandlerFunc>
+) => {
+    const dispatch = useDispatch();
+    useEffect(() => {
+        windowMessageHandler.current = (e: MessageEvent) => {
+            if (e.origin !== applicationUrl) return;
+            if (e.data?.type !== AUTH_RESULT_TYPE) return;
+            const { code, verifier } = e.data ?? {};
+            if (!code || !verifier) return;
+            localStorage.removeItem(LS_SIGN_IN_POPUP);
+            dispatch(tokenHandoff({ code, verifier, clientId: clientId, authCallbackUrl: authCallbackUrl }));
+        };
+        window.addEventListener("message", windowMessageHandler.current);
+
+        // Listener cleanup
+        return () => {
+            if (windowMessageHandler.current) {
+                window.removeEventListener("message", windowMessageHandler.current);
+            }   
+        };
+    }, [dispatch, applicationUrl, authCallbackUrl, clientId]);
+};
+
+export const useSessionWorkerTokenRefresh = (
+    clientId: string,
+    sessionWorkerRef: MutableRefObject<null | Worker>,
+    createWorker: () => Worker,
+    fetchUserDependentData: (servicesCb: () => void) => void,
```

Comment:
type signature question: shouldn't this be an async action type, im not sure if one exists in redux thunk? also since you don't pass the servicesCb in (instead it's pre-bound when used in bento_web), I think the signature is wrong anyway